tirocks/src/db/imp.rs

```diff
@@ -0,0 +1,726 @@
+        }
+    }
+
+    /// ListColumnFamilies will open the DB specified by argument name and return the list of all
+    /// column families in that DB. The ordering of returned column families is unspecified.
+    pub fn list_cfs(db: DbOptions, path: impl AsRef<Path>) -> Result<Vec<String>> {
+        let mut convert_s = Status::default();
+        let mut names = Vec::new();
+        unsafe {
+            let mut handle = |name: &[u8]| match str::from_utf8(name) {
+                Ok(n) => names.push(n.to_string()),
+                Err(e) => {
+                    convert_s = Status::with_error(Code::kCorruption, format!("{}", e));
+                }
+            };
+            let (ctx, fp) = util::wrap_string_receiver(&mut handle);
+            ffi_call!(crocksdb_list_column_families(
+                db.get_ptr(),
+                path.as_ref().path_to_slice(),
+                ctx,
+                Some(fp),
+            ))?;
+        }
+        if !convert_s.ok() {
+            return Err(convert_s);
+        }
+        Ok(names)
+    }
+
+    /// Create a column_family.
+    ///
+    /// If it's a titan db, it's the same as calling `create_cf_titan` with the titan specific
+    /// options set to default.
+    pub fn create_cf(&mut self, name: impl AsRef<str>, opt: CfOptions) -> Result<()> {
+        if self.is_titan {
+            return self.create_cf_titan(name, opt.into());
+        }
+        let ptr = unsafe {
+            ffi_call!(crocksdb_create_column_family(
+                self.ptr,
+                opt.as_ptr(),
+                r(name.as_ref().as_bytes()),
+            ))
+        }?;
+        if let Some(c) = opt.comparator() {
+            self.comparator.push(c.clone());
+        }
+        self.handles
+            .push(unsafe { RefCountedCfHandle::from_ptr(ptr, true) });
+        Ok(())
+    }
+
+    /// Create a column_family.
+    ///
+    /// If it's not titan, it's the same as calling `create_cf`. Titan specific options will be
+    /// ignored.
+    pub fn create_cf_titan(
+        &mut self,
+        name: impl AsRef<str>,
+        mut opt: TitanCfOptions,
+    ) -> Result<()> {
+        if !self.is_titan {
+            return self.create_cf(name, opt.into());
+        }
+        let ptr = unsafe {
+            ffi_call!(ctitandb_create_column_family(
+                self.ptr,
+                opt.as_mut_ptr(),
+                r(name.as_ref().as_bytes()),
+            ))
+        }?;
+        if let Some(c) = opt.comparator() {
+            self.comparator.push(c.clone());
+        }
+        self.handles
+            .push(unsafe { RefCountedCfHandle::from_ptr(ptr, true) });
+        Ok(())
+    }
+
+    /// Destroy a column family specified by name. This call records a drop record in the
+    /// manifest and prevents the column family from flushing and compacting.
+    pub fn destroy_cf(&mut self, name: &str) -> Result<bool> {
+        if name == DEFAULT_CF_NAME {
```

Comment:
There's already check in C++ code, no need to repeat it.

Reply:
No they are different. The default ColumnFamilyHandle should not be deleted by Rust code, it can only be deleted by rocksdb itself. And there is no check when destroying the handle.

Reply:
What do you mean? This is a detail only RocksDB code should care about. And RocksDB disallows dropping default cf.

Reply:
No, it's not an implementation details.

https://github.com/facebook/rocksdb/blob/490fcac07832205cb0bdb7ade24118d6567544fa/db/c.cc#L1055-L1058

If we don't have any protection, this will crash the program. On the other hand, if we are sure it should not be deleted, prevent it ahead of time make things simple instead of recovering from failure.

Reply:
I think there's a misunderstanding somewhere. You can't `destroy_cf`/`DropColumnFamily` a default cf in RocksDB (in default implementation). (https://github.com/tikv/rocksdb/blob/0b6d58bf0ef7529503cab8dcf9b9ada0b4cf2346/db/db_impl/db_impl.cc#L2194). The function you quote (`drop`/`DestroyColumnFamily`) is fine because each column family returned from RocksDB is a newly created pointer that can be freed on client's end.

Reply:
Never mind, the issue is fixed by `RefCountedCfHandler.managed`. The check here is just ahead of time check that avoids error recovery.

Reply:
There's not really any error recovery involved. I don't like doing these checks in a binding library, because you alter the original error text (and error type) so that a user will have to look at binding code to figure out how to deal with this error.

Reply:
For example, I need to check the returned value of `destroy`, and be sure if I need to re-push the cf back to the array. Otherwise user may be confused that you say `default cf` should not be deleted, but why I can't use default cf any more?

> because you alter the original error text (and error type) so that a user will have to look at binding code to figure out how to deal with this error

No, the error type is still the same. The text is not the same but similar: "default cf can't be dropped." vs "Can't drop default column family". And I don't see why user would need to look at the binding code to interpret the error.

Reply:
Well, that is really because you choose to `maybe_drop` the handle when RocksDB side doesn't associate `Destroy` with `Drop`. To me you should keep the handle (so that even a `DropColumnFamily`/`destroy_cf` succeeds, user can still get the handle via `DB::cf` and `DB::list_cfs`). This behavior is more consistent with RocksDB and doesn't require you to do error handling.

Reply:
I don't see any problem here. If it can make you feel better, I can change the error text exact the same as RocksDB internal.

> This behavior

The design of RocksDB's original API is to delay dropping when there is still a reference to the handle in user code. That is the exact behavior of `CfOptions`. Querying cf handle is a new design from Rust side to make it convenient for users that don't want to maintain their cf handle lifetime. I clean it up on purpose so I don't need to take care of other corner cases like recreating a column family with the same name.

Reply:
Let me rephrase. In RocksDB, a column family handle that is destroyed (via `DestroyColumnFamily`) can still be used to do a lot of things. So in this binding library, it is consistent to allow user to acquire and use that handle via `CfHandle::new(...)`. This means you shouldn't `swap_remove` and `maybe_drop` in this context.

After this behavior is applied, then there's no need to worry about handling the error of `DestroyColumnFamily`, and I don't see the benefits of checking the default cf in rust side.

BTW, only checking default cf is not enough for your original intention. There're probably plenty of places in RocksDB implementation that can error out. In that case user cannot recover gracefully either.

Reply:
> a column family handle that is dropped (via DropColumnFamily) can still be used to do a lot of things.

Yes, and the column family handle is returned when open the db. To do similar thing, user just need to do `CfHandle::new()` before dropping the column family. *As long as user holds a reference to column handle, dropping will not actually delete any data*. The statement remains the true.

>  There're probably plenty of places in RocksDB implementation that can error out.

Yes, I will comment that in the API. Since we are wrapping our fork and titan, it should not be a real problem.

Reply:
> The statement remains the true.

It's not the same statement. In RocksDB, the lifetime of a cf_handle starts in `CreateColumnFamily`, ends in `DestroyColumnFamilyHandle`. Calls to `DropColumnFamily` doesn't interfere with its lifetime.

In a practical sense, this is not friendly to user. Say a user calls `destroy_cf` in a separate thread, and do DB reads in another thread. And the read thread always acquires a cf on demand. In current implementation, it's not easy to coordinate the two threads.

Reply:
> not easy to coordinate the two threads.

If the cf is acquired before db is dropped, then the thread can still finish the read, otherwise it can't start reading.

Reply:
Yes, so it's not consistent with RocksDB. Also it occurs to me that `destroy_cf` requires `&mut self`. This is not good either. It's often impossible to acquire a mutable reference after DB is already shared in many places.

Reply:
It's already inconsistent with RocksDB already as you can't query any cf handle from RocksDB. The only thing in this case that can keep consistent is "As long as user holds a reference to column handle, dropping will not actually delete any data."

> destroy_cf requires &mut self.

I'm afraid this is a must because we maintain the cf list in db, it has to be cleaned up either when being destroyed or being dropped. And `DbRef` is implemented for `Arc<Mutex<DB>>` because of this exact case.

Reply:
But only this one requires a rust mutable reference (because of the `swap_remove`). Other places we only requires unsafe RocksDB reference which is transparent for rust users.

Reply:
I don't see your point. I already explain all these. If you insist we should stick with RocksDB style, I can also change it to return all the cf handles when open it instead of maintaining them in Db. Users can always wrap what they want base on it.

Reply:
There are two things that I'm concerned:

Different than RocksDB, tirocks manages the cf_handle, not the user. When user creates a cf, DB maintains the cf_handle so that user can always use `CfHandle::new` to get it. So intuitively, the cf_handle is managed by a DB instance, **user doesn't have to worry about keeping a handle himself because the DB instance is trusted with that responsibility**. User still have some control over the lifetime: the cf_handle is "alive" after `create_cf`/`CreateColumnFamily` (This behavior is consistent with RocksDB) and cf_handle can be lazily deleted after `destroy_cf` (**This is not consistent with RocksDB**).

What concerns me here is that you give away the responsibility of managing cf_handle to user in a call that shouldn't affect how user thinks of cf_handle. The example with reading thread is just demonstrating that user have to think about his responsibility in a wrong way (that is not reflected in the RocksDB interface).

Another thing is, a DB instance can be, and should be shared to multiple threads in a process. `DropColumnFamily`/`destroy_cf` can also be called concurrently by the users. But in your implementation, it cannot happen (even with a mutex, you are still imposing a synchronziation that is not necessary to RocksDB code).

Reply:
To be more specific, I'm not against managing the cf_handle by the DB. What I want to see is `destroy_cf` not deleting the DB owned cf_handle. And cf_handle is always available via `CfHandle::new` until DB is dropped.

Reply:
I don't think it's correct to keep the reference until DB is dropped. If a user wants to drop cf while reading, they probably recreate new column families. Keeping all the handles will OOM in the end. If a column family with the same names is recreated, it's impossible to correctly serve the cf query.

I don't think this tiny issue worth such long discussion. We should either stick with current implementation and well documented or just keep the RocksDB style that returns all the CF handles during open.

Reply:
That's true. How about locking these `handles` and `comparators` inside `DB`? At this point I just can't stand that user have to lock (and it has to be RWlock to have a decent performance) their DB to use `cf` operations after DB is initialized (and shared).

Reply:
Locking handles will make querying the handle expansive. The current design is well optimized for the use cases:
- Not sharing DB, there will be no allocations and no ref count;
- Sharing DB, but initialize cf before sharing it, there will be no locks;

The following use cases will have poor performance:
- Too many cfs, because querying cf is O(n), it can have poor performance;
- Sharing DB, and require dynamic creating/dropping cf, it requires lock.

Reply:
Fair enough.